theano/gpuarray/opt.py

```diff
@@ -895,7 +903,34 @@ def local_gpua_subtensor(op, context_name, inputs, outputs):
                                               for v in n.inputs + n.outputs])
                         for n, _ in clients[0]]):
                     return
+    # Here is the condition for the GraphToGPU opt. inputs is the
+    # inputs we want to use for the new node
+    if (x.owner and isinstance(x.owner.op, GpuFromHost)):
+        cpu_x = x.owner.inputs[0]
+        # And it is a shared var or an input of the graph.
+        # and is used by only 1 node.
+        # x is in the new graph, so we can't tests its number of clients.
```

Comment:
The input of the subtensor is always bigger then the output. So we try to be optimal. If it don't speed up computation to move the big input, then we keep the subtensor on the CPU and move to the GPU the smaller output.

This version of the condition is simpler to understand. I think they are mostly equivalent, but probably not strictly equivalent. What it tell, if we have HostFromGpu(x) and x is an input of the graph, and x have only 1 client (so only the HostFromGpu), then we are sure we don't need the full input to the GPU, so we transfer only the small output.

If we also need the big input as well as the small output on the GPU, then we will execute the subtensor on the GPU to only move the big input and not both on the GPU.